File: suites/04_compression_kernels/comparisons/huffman_coding.py
def huffman_checksum(data_size: int) -> int:
    checksum, total_bits = 0, 0
    for j in range(data_size):
        byte_val = (j * 7 + 13) % 256
        depth, val = 1, byte_val
        while val > 1:
            val //= 2
            depth += 1
        code_len = depth
        if code_len < 2: code_len = 2
        if code_len > 12: code_len = 12
        total_bits += code_len
        code = byte_val % (1 << code_len)
        checksum = (checksum * 37 + code * 19 + code_len * 7) % 1000000007
    return (checksum * 53 + total_bits) % 1000000007


if __name__ == "__main__":
    print(huffman_checksum(4096))

File: suites/04_compression_kernels/comparisons/huffman_coding.go
package main

import "fmt"

func huffmanChecksum(dataSize int64) int64 {
	checksum, totalBits := int64(0), int64(0)
	for j := int64(0); j < dataSize; j++ {
		byteVal := (j*7 + 13) % 256
		depth, val := int64(1), byteVal
		for val > 1 { val /= 2; depth++ }
		codeLen := depth
		if codeLen < 2 { codeLen = 2 }
		if codeLen > 12 { codeLen = 12 }
		totalBits += codeLen
		code := byteVal % (int64(1) << codeLen)
		checksum = (checksum*37 + code*19 + codeLen*7) % 1000000007
	}
	return (checksum*53 + totalBits) % 1000000007
}

func main() { fmt.Println(huffmanChecksum(4096)) }
